spec/primitives/float_spec.cr

```diff
@@ -27,8 +26,30 @@ describe "Primitives: Float" do
             (-{{ float }}::INFINITY).{{ method }}
           end
         end
+
+        it "raises overflow if not a number (#10421)" do
+          expect_raises(OverflowError) do
+            {{ float }}::NAN.{{ method }}
+          end
+        end
       {% end %}
     {% end %}
+
+    it "raises overflow if equal to Int::MAX (#11105)" do
+      # these examples hold because the integer would be rounded _up_ to the
+      # nearest representable float
+
+      expect_raises(OverflowError) { Float32.new!(Int32::MAX).to_i32 }
+      expect_raises(OverflowError) { Float32.new!(UInt32::MAX).to_u32 }
+      expect_raises(OverflowError) { Float32.new!(Int64::MAX).to_i64 }
+      expect_raises(OverflowError) { Float32.new!(UInt64::MAX).to_u64 }
+      expect_raises(OverflowError) { Float32.new!(Int128::MAX).to_i128 }
+
+      expect_raises(OverflowError) { Float64.new!(Int64::MAX).to_i64 }
+      expect_raises(OverflowError) { Float64.new!(UInt64::MAX).to_u64 }
+      expect_raises(OverflowError) { Float64.new!(Int128::MAX).to_i128 }
+      expect_raises(OverflowError) { Float64.new!(UInt128::MAX).to_u128 }
+    end
```

Comment:
This spec's definition is kinda circular.
Also would be nice to test that it *doesn't* overflow *right below* the cutoff.
I'm not sure if my suggestion is better, let me know what you think.

```suggestion
    describe "handles overflow near Int::MAX (#11105)", focus: true do
      describe "for Float32" do
        it "doesn't overflow below" do
          (n = 2.1474835e+9_f32).to_i32.should eq(n)
          (n = 4.294967e+9_f32).to_u32.should eq(n)
          (n = 9.2233715e+18_f32).to_i64.should eq(n)
          (n = 1.8446743e+19_f32).to_u64.should eq(n)
          (n = 1.7014117e+38_f32).to_i128.should eq(n)
          (n = Float32::MAX).to_u128.should eq(n)
        end

        it "overflows above" do
          expect_raises(OverflowError) { 2.1474836e+9_f32.to_i32 }
          expect_raises(OverflowError) { 4.2949673e+9_f32.to_u32 }
          expect_raises(OverflowError) { 9.223372e+18_f32.to_i64 }
          expect_raises(OverflowError) { 1.8446744e+19_f32.to_u64 }
          expect_raises(OverflowError) { 1.7014118e+38_f32.to_i128 }
        end
      end

      describe "for Float64" do
        it "doesn't overflow below" do
          (n = 9.223372036854775e+18).to_i64.should eq(n)
          (n = 1.844674407370955e+19).to_u64.should eq(n)
          (n = 1.7014118346046921e+38).to_i128.should eq(n)
          (n = 3.4028236692093843e+38).to_u128.should eq(n)
        end

        it "overflows above" do
          expect_raises(OverflowError) { 9.223372036854776e+18.to_i64 }
          expect_raises(OverflowError) { 1.8446744073709552e+19.to_u64 }
          expect_raises(OverflowError) { 1.7014118346046923e+38.to_i128 }
          expect_raises(OverflowError) { 3.402823669209385e+38.to_u128 }
        end
      end
    end
```

Reply:
I think you resolved one of my two concerns. But, I do think the idea of writing as exact number literals should be seriously considered. Because, if both the `Float32.new!` part and the `.to_u64` part break at the same time, it's still possible for this spec to pass. I will not push very hard on this, but curious to see some feedback.